src/org/rust/lang/core/resolve/RustResolveEngine.kt

```diff
@@ -80,6 +83,16 @@ public class RustResolveEngine() {
 
         private fun found(elem: RustNamedElement) {
             matched = elem
+
+            if (elem is RustModDeclItem) {
+                val mod = elem.parentModInFile ?: return
+                val name = elem.name ?: return
+                val path = mod.submodulePath(name)
+                if (path is ChildModFile.Found) {
+                    matched =  path.mod
+                }
+            }
```

Comment:
This ideally should be elsewhere, but I don't know where exactly. The similar logic can be used for `use foo::bar as baz` declarations. However `use foo::bar::*` seems completely different.